lib/datadog/di.rb

```diff
@@ -10,5 +16,42 @@ module Datadog
   module DI
     # Expose DI to global shared objects
     Extensions.activate!
+
+    class << self
+      attr_reader :code_tracker
+
+      # Activates code tracking. Normally this method should be called
+      # when the application starts. If instrumenting third-party code,
+      # code tracking needs to be enabled before the third-party libraries
+      # are loaded. If you definitely will not be instrumenting
+      # third-party libraries, activating tracking after third-party libraries
+      # have been loaded may improve lookup performance.
+      #
+      # TODO test that activating tracker multiple times preserves
+      # existing mappings in the registry
+      def activate_tracking!
+        @code_tracker ||= CodeTracker.new
+        # & is demanded by steep, code tracker is always not nil here.
+        code_tracker&.start
+      end
+
+      # Deactivates code tracking. In normal usage of DI this method should
+      # never be called, however it is used by DI's test suite to reset
+      # state for individual tests.
+      #
+      # Note that deactivating tracking clears out the registry, losing
+      # the ability to look up files that have been loaded into the process
+      # already.
+      def deactivate_tracking!
+        code_tracker&.stop
+      end
+
+      # Returns whether code tracking is available.
+      # This method should be used instead of querying #code_tracker
+      # because the latter one may be nil.
+      def code_tracking_active?
+        code_tracker&.active? || false
+      end
```

Comment:
I think steep complains because all 3 methods are order dependent. Without calling `activate_tracking!` instance variable will not be set. And because of that it could be `nil`.

Maybe we can follow common practice with reader like this

```ruby
def code_tracker
  @code_tracker ||= CodeTracker.new
end
```

I think after that Steep should be quite (and it will allow to untangle the order dependency)

Reply:
Code tracker should only exist if code tracking is active. Checking whether code tracker exists shouldn't create it.